bolt/src/Passes/AdrPass.cpp

```diff
@@ -0,0 +1,73 @@
+//===--- AdrPass.cpp ------------------------------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//===----------------------------------------------------------------------===//
+
+#include "AdrPass.h"
+#include "ParallelUtilities.h"
+
+using namespace llvm;
+
+namespace opts {
+extern cl::OptionCategory BoltCategory;
+
+static cl::opt<bool>
+    AdrPassOpt("adrpass-opt",
+               cl::desc("Replace ARM non-local ADR instructions with ADRP"),
+               cl::init(true), cl::cat(BoltCategory), cl::ReallyHidden);
+} // namespace opts
+
+namespace llvm {
+namespace bolt {
+
+void AdrPass::runOnFunction(BinaryContext &BC, BinaryFunction &BF) {
+  for (BinaryBasicBlock *BB : BF.layout()) {
+    for (auto It = BB->begin(); It != BB->end(); ++It) {
+      MCInst &Inst = *It;
+      if (!BC.MIB->isADR(Inst))
+        continue;
+
+      uint64_t Target;
+      if (BC.MIB->evaluateMemOperandTarget(Inst, Target))
+        continue;
+
+      if (!BC.MIB->hasAnnotation(Inst, "Offset"))
+        continue;
+
+      uint32_t Offset = BC.MIB->getAnnotationAs<uint32_t>(Inst, "Offset");
+      BC.MIB->removeAnnotation(Inst, "Offset");
+
+      const Relocation *Relocation = BF.getRelocationAt(Offset);
+      if (!Relocation)
+        continue;
+
+      MCPhysReg Reg;
+      BC.MIB->getADRReg(Inst, Reg);
+      std::vector<MCInst> Addr = BC.MIB->getAddress(
+          Relocation->Symbol, BC.Ctx.get(), Reg, Relocation->Addend);
+      It = BB->replaceInstruction(It, Addr.begin(), Addr.end());
+    }
```

Comment:
I was thinking about this solution, but it will over complicate the code as it seemed to me. We will need to get symbol, addend from the current instruction, create newer ones with getAddress(), found the ADRP instruction and create new expression for it.. Maybe I'm wrong and saving offsets does not worth it, although I was thinking the ADR instruction is quite rare one and it is OK to do it like this.. I will try to figure out how to re-use information from the ADR instruction it self ,maybe it won't be to complicated to do..